leader-election-impl/src/main/java/com/palantir/leader/PaxosLeadershipEventRecorder.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.leader;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import com.codahale.metrics.MetricRegistry;
+import com.google.common.annotations.VisibleForTesting;
+import com.palantir.paxos.PaxosRoundFailureException;
+import com.palantir.paxos.PaxosValue;
+
+public class PaxosLeadershipEventRecorder implements PaxosKnowledgeEventRecorder, PaxosLeaderElectionEventRecorder {
+
+    private final String leaderId;
+    private final LeadershipEvents events;
+
+    @GuardedBy("this") private PaxosValue currentRound = null;
+    @GuardedBy("this") private boolean isLeading = false;
+
+    public static PaxosLeadershipEventRecorder create(MetricRegistry metrics, String leaderUuid) {
+        return new PaxosLeadershipEventRecorder(new LeadershipEvents(metrics), leaderUuid);
+    }
+
+    @VisibleForTesting
+    PaxosLeadershipEventRecorder(LeadershipEvents events, String leaderUuid) {
+        this.events = events;
+        this.leaderId = leaderUuid;
+    }
+
+    @Override
+    public void recordProposalAttempt(long round) {
+        events.proposedLeadershipFor(round);
+    }
+
+    @Override
+    public void recordProposalFailure(PaxosRoundFailureException e) {
+        events.proposalFailure(e);
+    }
+
+    @Override
+    public synchronized void recordRound(PaxosValue round) {
+        if (isNewRound(round)) {
+            recordNewRound(round);
+        }
+    }
+
+    private void recordNewRound(PaxosValue round) {
+        if (isLeading) {
+            events.lostLeadershipFor(currentRound);
+        }
+
+        if (isLeaderFor(round)) {
+            events.gainedLeadershipFor(round);
+        }
+
+        currentRound = round;
+        isLeading = isLeaderFor(round);
+    }
+
+    @Override
+    public synchronized void recordNotLeading(PaxosValue value) {
+        if (isSameRound(value)) {
+            if (isLeading) {
+                events.lostLeadershipFor(value);
+                isLeading = false;
+            }
+        }
+    }
+
+    @Override
+    public synchronized void recordNoQuorum(PaxosValue value) {
+        if (isSameRound(value)) {
+            events.noQuorum(value);
+        }
+    }
+
+    private boolean isNewRound(PaxosValue value) {
+        return value != null && (currentRound == null || value.getRound() > currentRound.getRound());
```

Comment:
Yeah, this is something I thought about; my conclusions were that (a) we don't want the metrics to show a gainedLeadership after a lostLeadership in this case, since that would imply we're currently the leader, and (b) we didn't gain leadership in a practical sense, since we lost it before we knew we had it. In any case, will definitely add a test for this!

There probably is value in logging every paxos round regardless of order - if you agree I'll make an issue for that.